File: uvm/hier/usr/lib/python2.5/sql_helper.py
import logging
import psycopg
import re

from datetime import date
from datetime import timedelta
from sets import Set

# XXX function timing

__conn = None

def get_connection():
    global __conn

    if not __conn:
        __conn = psycopg.connect("dbname=uvm user=postgres")

    return __conn

def create_table_from_query(tablename, query, args=None):
    drop_table(tablename)
    create_table_as_sql(tablename, query, args)

def drop_table(tablename):
    conn = get_connection()

    try:
        curs = conn.cursor()
        curs.execute("DROP TABLE %s" % tablename)
    except:
        logger.debug('did not drop table: %s' % tablename)
    finally:
        conn.commit()

def create_table_as_sql(tablename, query, args):
    run_sql("CREATE TABLE %s AS %s" % (tablename, query), args)

def run_sql(sql, args=None):
    conn = get_connection()

    try:
        curs = conn.cursor()
        curs.execute(sql, args)
    finally:
        conn.commit()

def create_partitioned_table(table_ddl, timestamp_column, start_date, end_date,
                             clear_tables=False):
    (schema, tablename) = __get_tablename(table_ddl)

    existing_dates = Set()

    for t in get_tables(schema='reports', prefix='%s_' % tablename):
        m=re.search('%s_(\d+)_(\d+)_(\d+)' % tablename, t)
        if m:
            d = date(*map(int, m.groups()))
            if d >= start_date and d < end_date:
                existing_dates.add(d)
            else:
                drop_table(t, schema='reports')
        else:
            logging.warn('ignoring table: %s' % tablename)

    interval = (end_date - start_date).days

    all_dates = Set(end_date - timedelta(days=i + 1) for i in range(interval))

    for d in all_dates - existing_dates:
        run_sql("""\
CREATE TABLE %s
(CHECK (%s >= %%s AND %s < %%s))
INHERITS (%s)""" % (__tablename_for_date(tablename, d),
                    timestamp_column, timestamp_column, tablename),
        (d, d + timedelta(days=1)))

    if clear_tables:
        for d in all_dates:
            drop_table(__tablename_for_date(tablename, d))

    __make_trigger(tablename, all_dates)

def get_update_info(tablename):
    conn = get_connection()
    try:
        curs = conn.cursor()

        curs.execute("""\
SELECT last_update FROM reports.table_updates WHERE tablename = %s
""", (tablename,))

        row = curs.fetchone()

        if row:
            rv = row[0]
        else:
            rv = None

    finally:
        conn.commit()

    return rv

def set_update_info(tablename, last_update):
    conn = get_connection()
    try:
        curs = conn.cursor()

        curs.execute("""\
SELECT count(*) FROM reports.table_updates WHERE table_updates = %s
""", (tablename,))
        row = curs.fetchone()

        if row[0] == 0:
            curs = conn.cursor()
            curs.execute("""\
INSERT INTO reports.table_updates (tablename, last_update) VALUES (%s, %s)
""", (tablename, date))
        else:
            curs = conn.cursor()
            curs.execute("""\
UPDATE reports.table_updates SET last_update = %s WHERE tablename = %s
""", (date, tablename))
    finally:
        conn.commit();

def drop_table(table, schema=None):
    if schema:
        tn = '%s.%s' % (schema, table)
    else:
        tn = table

    conn = get_connection()
    try:
        curs = conn.cursor()
        curs.execute('DROP TABLE %s' % tn)
    except psycopg.ProgrammingError:
        logging.debug('cannot drop table: %s' % table)
    finally:
        conn.commit()

def get_tables(schema=None, prefix=''):
    conn = get_connection()

    try:
        curs = conn.cursor()

        if schema:
            curs.execute("""\
SELECT tablename FROM pg_catalog.pg_tables
WHERE schemaname = %s AND tablename LIKE %s""", (schema, '%s%%' % prefix))
        else:
            curs.execute("""\
SELECT tablename FROM pg_catalog.pg_tables
WHERE tablename LIKE %s""", '%s%%' % prefix)

        rows = curs.fetchall()
        rv = [rows[i][0] for i in range(len(rows))]
    finally:
        conn.commit()

    return rv

def __make_trigger(tablename, all_dates):
    trigger_function_name = '%s_insert_trigger()' % tablename

    trigger_function = """\
CREATE OR REPLACE FUNCTION %s
RETURNS TRIGGER AS $$
BEGIN
""" % trigger_function_name

    first = True
    for d in all_dates:
        trigger_function += """\
    %s (NEW.%s >= '%s' AND NEW.%s < '%s') THEN
        INSERT INTO %s VALUES (NEW.*);""" % ('IF' if first else "ELSIF",
                                             timestamp_column, d,
                                             timestamp_column,
                                             d + timedelta(days=1),
                                             __get_tablename(tablename, d))
        first = False

    trigger_function += """\
    ELSE
        RAISE NOTICE 'Date out of range: %', NEW.#{timestamp_column};
    END IF;
    RETURN NULL;
END;
$$
LANGUAGE plpgsql;"""

    run_sql(trigger_function);

    trigger_name = "insert_%s_trigger" % tablename

    if not __has_trigger(schema, tablename, trigger_name % tablename):
        run_sql("""\
CREATE TRIGGER trigger_name
    BEFORE INSERT ON %s
    FOR EACH ROW EXECUTE PROCEDURE %s
""" % (trigger_name, tablename, trigger_function_name))

def __tablename_for_date(tablename, date):
    return "%s_%d_%d_%d" % ((tablename,) + date.timetuple()[0:3])

def __get_tablename(table_ddl):
    m = re.search('create\s+table\s+(\S+)', table_ddl, re.I | re.M)

    if m:
      s = m.group(1).split('.')
      return (s[0:-1], s[-1])
    else:
      raise ValueError("Cannot find table in: %s" % table_ddl)

File: uvm-lib/hier/usr/lib/ruby/1.8/sql-helper.rb
require 'set'

class SqlHelper
  def SqlHelper.log_sql_error(msg, e)
    LOG.warn <<MSG
#{msg}
Error code: #{e.err}
Error message: #{e.errstr}
Error SQLSTATE: #{e.state}
MSG
  end

  def initialize(dbh)
    @dbh = dbh
  end

  def next_id()
    @dbh.select_one(<<SQL).first
SELECT nextval('hibernate_sequence')
SQL
  end

  def add_column(table, column, type, options)
    log_sql(<<SQL)
ALTER TABLE #{table} ADD COLUMN #{column} #{type}
SQL
    yield

    if options[:not_null]
      log_sql(<<SQL)
ALTER TABLE #{table} ALTER COLUMN #{column} SET NOT NULL
SQL
    end
  end

  def remove_columns(table, columns, cascade = true)
    [columns].flatten.each do |c|
      begin
      log_sql(<<SQL)
ALTER TABLE #{table} DROP COLUMN #{c}#{' CASCADE' if cascade}
SQL
      rescue DBI::DatabaseError => e
        SqlHelper.log_sql_error("Could not remove column #{c}", e)
      end
    end
  end

  def rename_column(table, old_name, new_name)
    log_sql(<<SQL)
ALTER TABLE #{table} RENAME COLUMN #{old_name} TO #{new_name}
SQL
  end

  def drop_tables(tables, cascade = true)
    begin
      log_sql(<<SQL)
DROP TABLE #{[tables].flatten.join(', ')}#{' CASCADE' if cascade}
SQL
    rescue DBI::DatabaseError => e
      SqlHelper.log_sql_error("Could not remove tables #{[tables].flatten.join(', ')}", e)
    end
  end

  def has_trigger?(schema, table, trigger_name)
    @dbh.prepare(<<SQL) do |ps|
SELECT 1 FROM information_schema.triggers
WHERE trigger_schema = ? AND event_object_table = ? AND trigger_name = ?
SQL
      ps.execute(schema, table, trigger_name)
      l = ps.fetch()
      !l.nil?
    end
  end

  def get_col_type(schema, table, column)
    @dbh.prepare(<<SQL) do |ps|
SELECT data_type
FROM information_schema.columns
WHERE table_schema = ? AND table_name = ? AND column_name = ?
SQL
      ps.execute(schema, table, column)
      l = ps.fetch()
      l.nil? ? nil : l[0]
    end
  end

  def synchronize_table(table_name, primary_key, list)
    keys = get_keys(primary_key, list)
    bad_keys = [ ]
    update_row_keys = Set.new

    @dbh.execute("SELECT #{primary_key.to_s} FROM #{table_name}") do |r|
      r.fetch do |e|
        k = e.first
        if keys.member?(k)
          update_row_keys << k
        else
          bad_keys << k
        end
      end
    end

    new_row_keys = keys - update_row_keys.to_a
    new_rows = list.select { |e| new_row_keys.member?(e[primary_key]) }
    insert_rows(table_name, new_rows)

    update_rows = list.select { |e| update_row_keys.member?(e[primary_key]) }
    update_rows(table_name, primary_key, update_rows)

    @dbh.prepare("DELETE FROM #{table_name} WHERE #{primary_key} = ?") do |ps|
      bad_keys.each do |e|
        ps.execute(e)
      end
    end
  end

  def create_table_from_query(table, sql, *params)
    begin
      log_sql("DROP TABLE #{table}")
    rescue
      # fine if no table
    end

    begin
      @dbh.prepare("CREATE TABLE #{table} AS #{sql}") do |ps|
        ps.execute(*params)
      end
    rescue DBI::DatabaseError => e
      SqlHelper.log_sql_error("Could not create table from #{sql} (#{params.join(', ')})", e)
    end
  end

  def get_short_table_name(table)
    /\.(.*)$/ =~ table ? $1 : table
  end

  def get_schema(table)
    /^([^.]+)\./ =~ table ? $1 : 'public'
  end

  def table_exist?(table)
    @dbh.tables.member?(get_short_table_name(table))
  end

  def do_update(name, end_date, interval=30)
    @dbh['AutoCommit'] = false
    @dbh.transaction do |dbh|
      last_update = @dbh.prepare(<<SQL) do |ps|
SELECT last_update FROM reports.updates WHERE name = ?
SQL
        ps.execute(name)
        l = ps.fetch()
        l.nil? ? nil : l[0]
      end

      if (last_update.nil?)
        yield(true, end_date - interval, end_date)
        @dbh.prepare(<<SQL) do |ps|
INSERT INTO reports.updates VALUES (?, ?)
SQL
          ps.execute(name, end_date)
        end
      else
        sd = if (end_date - interval > last_update.to_date)
               end_date - interval
             else
               last_update
             end
        yield(false, sd, end_date)
        @dbh.prepare(<<SQL) do |ps|
UPDATE reports.updates SET last_update = ? WHERE name = ?
SQL
          ps.execute(end_date, name)
        end
      end
    end
    @dbh['AutoCommit'] = true
  end

  def create_partitioned_table(table_ddl, timestamp_column,
                               end_date, interval=30)
    t0 = Time.now
    if interval <= 0
      raise "Bad interval: #{interval}"
    end

    start_date = end_date - interval;

    if /create\s+table\s+(\S+)/im =~ table_ddl
      table_name = $1
      short_table_name = get_short_table_name(table_name)
      schema = get_schema(table_name)
    else
      raise "Cannot find table in: #{table_ddl}"
    end

    dates = Set.new
    start_date.upto(end_date - 1) {|d| dates << d}

    @dbh['AutoCommit'] = false
    @dbh.transaction do |dbh|

      existing_dates = Set.new

      @dbh.do(table_ddl) unless table_exist?(table_name)

      @dbh.tables.each do |t|
        if /^#{short_table_name}_(.*)$/ =~ t
          suffix = $1
          date = Date.parse(suffix.gsub('_', '-'))
          if dates.member?(date)
            existing_dates << date
          else
            @dbh.do("DROP TABLE #{table_name}_#{suffix}")
          end
        end
      end

      query_start_date = nil

      (dates - existing_dates).each do |d|
        @dbh.prepare(<<SQL) do |ps|
CREATE TABLE #{table_name}_#{d.to_s.gsub('-', '_')}
(CHECK (#{timestamp_column} >= ? AND #{timestamp_column} < ?))
INHERITS (#{table_name})
SQL
          ps.execute(d, d.succ)
        end

        if query_start_date.nil? or query_start_date > d
          query_start_date = d
        end
      end

      trigger_function = <<SQL
CREATE OR REPLACE FUNCTION #{short_table_name}_insert_trigger()
RETURNS TRIGGER AS $$
BEGIN
SQL

      first = true
      dates.each do |d|
        trigger_function += <<SQL
    #{first ? 'IF' : 'ELSIF'} (NEW.#{timestamp_column} >= '#{d}'
          AND NEW.#{timestamp_column} < '#{d.succ}') THEN
        INSERT INTO #{table_name}_#{d.to_s.gsub('-', '_')} VALUES (NEW.*);
SQL
        first = false
      end

      trigger_function += <<SQL
    ELSE
        RAISE NOTICE 'Date out of range: %', NEW.#{timestamp_column};
    END IF;
    RETURN NULL;
END;
$$
LANGUAGE plpgsql;
SQL

      @dbh.do(trigger_function)

      unless has_trigger?(schema, short_table_name,
                          "insert_#{short_table_name}_trigger")
        @dbh.do <<SQL
CREATE TRIGGER insert_#{short_table_name}_trigger
    BEFORE INSERT ON #{table_name}
    FOR EACH ROW EXECUTE PROCEDURE #{short_table_name}_insert_trigger()
SQL
      end

      unless query_start_date.nil?
        query_start_date.upto(end_date - 1) do |d|
          @dbh.do("TRUNCATE TABLE #{table_name}_#{d.to_s.gsub('-', '_')}")
        end

        yield(query_start_date, end_date)
      end

    end
    @dbh['AutoCommit'] = true
    t1 = Time.now
    LOG.info "#{table_name} took: #{t1 - t0} seconds"
  end

  def log_sql(sql)
    t0 = Time.now
    @dbh.do(sql)
    t1 = Time.now

    LOG.debug "#{sql} took: #{t1 - t0} seconds"
  end

  private

  def get_keys(primary_key, list)
     list.inject(Set.new) { |m, o| m << o[primary_key] }.to_a
  end

  def get_all_columns(rows)
    rows.inject(Set.new) do |m, o|
      m += o.instance_of?(Hash) ? o.keys : o.field_names
    end.to_a
  end

  def bind_values(columns, row)
    columns.map { |c| row[c] }
  end

  def get_insert_statement(table, columns)
    <<SQL
INSERT INTO #{table} (#{columns.join(', ')})
VALUES (#{columns.map {|e| '?'}.join(', ') })
SQL
  end

  def get_update_statement(table, primary_key, columns)
    <<SQL
UPDATE #{table}
SET #{columns.map { |c| "#{c} = ?"}.join(', ')}
WHERE #{primary_key} = ?
SQL
  end

  def insert_rows(table, rows)
    columns = get_all_columns(rows)
    @dbh.prepare(get_insert_statement(table, columns)) do |ps|
      rows.each do |r|
        ps.execute(*bind_values(columns, r))
      end
    end
  end

  def update_rows(table, primary_key, rows)
    columns = get_all_columns(rows).reject { |c| primary_key == c }
    @dbh.prepare(get_update_statement(table, primary_key, columns)) do |ps|
      rows.each do |r|
        ps.execute(*bind_values(columns + [primary_key], r))
      end
    end
  end
end

